flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractHaServices.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.highavailability;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.blob.BlobStore;
+import org.apache.flink.runtime.blob.BlobStoreService;
+import org.apache.flink.runtime.leaderelection.LeaderElectionService;
+import org.apache.flink.runtime.leaderretrieval.LeaderRetrievalService;
+import org.apache.flink.util.ExceptionUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.Executor;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Abstract high availability services based on distributed system(e.g. Zookeeper, Kubernetes).
+ */
+public abstract class AbstractHaServices implements HighAvailabilityServices {
```

Comment:
Hmm. For `LeaderElectionService`, I agree with you that the `lock` should be dealt very carefully and it is not very easy to do in the sub classes. So I will factor the `LeaderElectionService` with composition.

However, for the `AbstractHaServices`, we do not run into this situation and the subclass is very easy to implement. Another concern is lots of codes will be touched if we rename the `ZookeeperHaServices` to `ZookeeperHaDriver` or sth similar. I have tried and find it is not worth.

Reply:
To avoid the renaming issue one could keep `ZooKeeperHaServices` as an alias for the `HaServices` with a `ZooKeeperDriver` (e.g. `ZooKeeperHaServices extend HaServices` and then in the constructor `super(new ZooKeeperDriver())`), for example. But I haven't checked it very thoroughly.

Reply:
This is an interesting solution. Unlike `LeaderElectionService`, we could not get too much benefit from this. So I will  leave this refactor in the future when we want to touch the `ZookeeperHaService` again. For example, get rid of the `lock-and-release` and use a K8s similar transactional operation.